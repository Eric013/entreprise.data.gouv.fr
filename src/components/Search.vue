<template>
  <div class="search light-text">
    <div class="container">
      <transition name="fade">
        <div class="search__text" v-if="showText">
          <h1 class="search__title">
            Retrouvez toutes les informations concernant les entreprises et associations de France
          </h1>
          <p class="search__subtitle">La plus grande base de données sur l'état civil des entreprises françaises est maintenant accessible à tous, sans frais.<br /><router-link to="/">Plus d'informations</router-link>.</p>
        </div>
      </transition>
      <SearchBar searchName="Recherche par nom"></SearchBar>
      <div class="filters">
        <button class="btn btn-default" id="filter-button" v-on:click="filterButton">
          Filtres
          <svg class="icon icon-equalizer"><use xlink:href="#icon-equalizer"></use></svg>
        </button>
        <SearchBarSmall searchName="Code Postal" v-if="toggleFilters"></SearchBarSmall>
        <SearchBarSmall searchName="Activite Principale" v-if="toggleFilters"></SearchBarSmall>
      </div>
    </div>
  </div>
</template>

<script>
import SearchBar from '@/components/searchbars/SearchBar.vue'
import SearchBarSmall from '@/components/searchbars/SearchBarSmall.vue'
import Results from '@/components/Results.vue'

export default {
  name: 'Search',
  components: {
    'SearchBar': SearchBar,
    'SearchBarSmall': SearchBarSmall,
    'Results': Results
  },
  data () {
    return {
      toggleFilters: true,
      results: null
    }
  },
  methods: {
    filterButton () {
      this.toggleFilters = !this.toggleFilters
      this.$store.commit('clearFilters')
      this.$store.commit('executeSearch')
    },
    requestSearch () {
      this.$store.commit('executeSearch')
    },
    clearButton () {
      this.$store.commit('clearResults')
    }
  },
  computed: {
    showText () {
      return this.$store.state.isSearchTextVisible
    }
  }
}

</script>

<style lang="scss" scoped>
  .search {
    padding: 3em;
    text-align: center;
    background: linear-gradient(180deg, $color-light-blue, $color-medium-blue);
    box-sizing: content-box;
  }

  .search__subtitle {
    margin: 2em 0;
  }

  .marged-top {
    margin-top: 1em;
  }

  #filter-button {
    margin: 1em;
  }

  .icon-equalizer {
    font-size: 23px;
  }

  .filters {
    display: none;
  }
</style>
