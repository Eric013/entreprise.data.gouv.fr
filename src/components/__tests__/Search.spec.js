import Search from '@/components/Search.vue'
import { createLocalVue, shallow, mount } from '@vue/test-utils'
import { __createMocks as createStoreMocks } from '@/store/index.js'
import Vuex from 'vuex'

const localVue = createLocalVue()
localVue.use(Vuex)
jest.mock('@/store/index.js')
jest.mock('@/router/index.js')

describe('Results.vue', () => {
  let storeMocks
  let wrapperSearch
  let search
  let $route

  beforeEach(() => {
    $route = {
      params: 'mock-params',
      push: jest.fn(),
      path: 'mock-path',
      query: {
        page: 2,
        fullText: 'mock-fullText'
      }
    }

    storeMocks = createStoreMocks()

    wrapperSearch = shallow(Search, {
      localVue,
      store: storeMocks.store,
      mocks: { $route },
      stubs: ['router-link', 'router-view']
    })
    search = wrapperSearch.vm
  })

  test('$route is watched, and will requestSearch if there is a fullText', () => {
    wrapperSearch.update(() => {
      expect(wrapperSearch.vm.$store.actions.requestSearch.mock.calls).toHaveLength(0)
    })
  })

  test('$route is watched, but wont requestSearch if there is not a fullText', () => {
    wrapperSearch = shallow(Search, {
      localVue,
      store: new Vuex.Store({
        state: storeMocks.state,
        getters: storeMocks.getters,
        mutations: {
          setPage: jest.fn(),
          setFullText: jest.fn()
        },
        actions: {
          requestSearch: jest.fn()
        }
      }),
      mocks: { $route },
      stubs: ['router-link', 'router-view']
    })
    wrapperSearch.update(() => {
      expect(wrapperSearch.vm.$store.actions.requestSearch.mock.calls).toHaveLength(0)
    })
  })

  test('Computed value isSearchNotEmpty check if search is not empty', () => {
    storeMocks.store.state.storedFullText = ''
    wrapperSearch.update()
    expect(wrapperSearch.vm.isSearchNotEmpty).toBeFalsy()
    storeMocks.store.state.storedFullText = 'placeholder'
    wrapperSearch.update()
    expect(wrapperSearch.vm.isSearchNotEmpty).toBeTruthy()
  })

  test('Computed value showWelcomeText works', () => {
    expect(search.showWelcomeText).toBe(search.$store.state.welcomeText.isWelcomeTextVisible)
  })

  test('Computed value showBackToResultsButton is truthy only on etablissement page, \
      if numberResults > 1, and storedStatus === 200', () => {
    $route = {
      params: 'mock-params',
      push: jest.fn(),
      path: '/entreprise/mock-searchId',
      query: {
        page: 2,
        fullText: 'mock-fullText'
      }
    }
    wrapperSearch = shallow(Search, {
      localVue,
      store: new Vuex.Store({
        state: {
          welcomeText: {
            isWelcomeTextVisible: true
          },
          results: {
            storedStatus: 200
          },
          search: {
            storedFullText: 'mock-storedFullText'
          }
        },
        mutations: {
          setPage: jest.fn(),
          setFullText: jest.fn()
        },
        actions: {
          requestSearch: jest.fn()
        },
        getters: {
          numberResults: jest.fn().mockReturnValue(2)
        },
      }),
      mocks: { $route },
      stubs: ['router-link', 'router-view']
    })
    expect(wrapperSearch.vm.showBackToResultsButton).toBeTruthy()
  })

  test('Computed value showBackToResultsButton is falsy if not on etablissement page, \
      if numberResults > 1, and storedStatus === 200', () => {
    $route = {
      params: 'mock-params',
      push: jest.fn(),
      path: '/results/mock-params',
      query: {
        page: 2,
        fullText: 'mock-fullText'
      }
    }
    wrapperSearch = shallow(Search, {
      localVue,
      store: new Vuex.Store({
        state: {
          welcomeText: {
            isWelcomeTextVisible: true
          },
          results: {
            storedStatus: 200
          },
          search: {
            storedFullText: 'mock-storedFullText'
          }
        },
        mutations: {
          setPage: jest.fn(),
          setFullText: jest.fn()
        },
        actions: {
          requestSearch: jest.fn()
        },
        getters: {
          numberResults: jest.fn().mockReturnValue(2)
        },
      }),
      mocks: { $route },
      stubs: ['router-link', 'router-view']
    })
    expect(wrapperSearch.vm.showBackToResultsButton).toBeFalsy()
  })

  test('Computed value showBackToResultsButton is falsy if on etablissement page, \
    if numberResults === 1, and storedStatus === 200', () => {
    $route = {
      params: 'mock-params',
      push: jest.fn(),
      path: '/results/mock-params',
      query: {
        page: 2,
        fullText: 'mock-fullText'
      }
    }
    wrapperSearch = shallow(Search, {
      localVue,
      store: new Vuex.Store({
        state: {
          welcomeText: {
            isWelcomeTextVisible: true
          },
          results: {
            storedStatus: 200
          },
          search: {
            storedFullText: 'mock-storedFullText'
          }
        },
        mutations: {
          setPage: jest.fn(),
          setFullText: jest.fn()
        },
        actions: {
          requestSearch: jest.fn()
        },
        getters: {
          numberResults: jest.fn().mockReturnValue(2)
        },
      }),
      mocks: { $route },
      stubs: ['router-link', 'router-view']
    })
    expect(wrapperSearch.vm.showBackToResultsButton).toBeFalsy()
  })
  test('Computed value showBackToResultsButton is falsy only on etablissement page, \
    if numberResults > 1, and storedStatus !== 200', () => {
    $route = {
      params: 'mock-params',
      push: jest.fn(),
      path: '/entreprise/mock-searchId',
      query: {
        page: 2,
        fullText: 'mock-fullText'
      }
    }
    wrapperSearch = shallow(Search, {
      localVue,
      store: new Vuex.Store({
        state: {
          welcomeText: {
            isWelcomeTextVisible: true
          },
          results: {
            storedStatus: 404
          },
          search: {
            storedFullText: 'mock-storedFullText'
          }
        },
        mutations: {
          setPage: jest.fn(),
          setFullText: jest.fn()
        },
        actions: {
          requestSearch: jest.fn()
        },
        getters: {
          numberResults: jest.fn().mockReturnValue(2)
        },
      }),
      mocks: { $route },
      stubs: ['router-link', 'router-view']
    })
    expect(wrapperSearch.vm.showBackToResultsButton).toBeFalsy()
  })

  test('At creation, if there is a query page, we set the correct page', () => {
    expect(storeMocks.mutations.setPage).toHaveBeenCalledWith(storeMocks.state, 2)
  })

  test('At creation, if there is fullText in route query, we set it and launch search', () => {
    expect(storeMocks.mutations.setFullText).toHaveBeenCalledWith(storeMocks.state, 'mock-fullText')
    expect(storeMocks.actions.requestSearch).toHaveBeenCalled()
  })

  // TODO: test when router-link will be testable
  test('Clicking back to results buttons goes back to result page')
})

describe('Search.vue snapshot', () => {
  const storeMocks = createStoreMocks()

  const $route = {
    params: 'mock-params',
    push: jest.fn(),
    path: 'mock-path',
    query: {
      page: 2,
      fullText: 'mock-fullText'
    }
  }

  const wrapperSearch = mount(Search, {
    localVue,
    store: storeMocks.store,
    mocks: { $route },
    stubs: ['router-link', 'router-view']
  })

  test('It match the snapshot when there arent results', () => {
    expect(wrapperSearch.vm.$el).toMatchSnapshot()
  })

  // TODO: test when router-link will be testable
  test('It match the snapshot when there are results')
})