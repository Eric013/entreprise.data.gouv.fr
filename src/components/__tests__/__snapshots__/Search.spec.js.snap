// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search.vue snapshot It match the snapshot 1`] = `
<div>
  <div
    class="hero"
  >
    <div
      class="notification full-width"
    >
      
          Ce site est un travail en cours, actuellement en beta. Vous pouvez le consulter librement.
        
    </div>
     
    <div
      class="hero__container container hero__compact"
    >
      <!---->
       
      <!---->
       
      <!---->
    </div>
  </div>
   
  <!---->
</div>
`;
